pkg/glusterfs/controllerserver.go

```diff
@@ -405,15 +454,243 @@ func (cs *ControllerServer) ControllerGetCapabilities(ctx context.Context, req *
 
 // CreateSnapshot create snapshot of an existing PV
 func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+
+	if err := cs.validateCreateSnapshotReq(req); err != nil {
+		return nil, err
+	}
+	glog.V(2).Infof("received request to create snapshot %v from volume %v", req.GetName(), req.GetSourceVolumeId())
+
+	snapInfo, err := cs.GfDriver.client.SnapshotInfo(req.Name)
+	if err != nil {
+		glog.Errorf("failed to get snapshot info for %v with Error %v", req.GetName(), err.Error())
+		errResp := cs.client.LastErrorResponse()
+		//errResp will be nil in case of No route to host error
+		if errResp != nil && errResp.StatusCode != http.StatusNotFound {
+
+			return nil, status.Errorf(codes.Internal, "CreateSnapshot - failed to get snapshot info %s", err.Error())
+		}
+		if errResp == nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+	} else {
+
+		if snapInfo.ParentVolName != req.GetSourceVolumeId() {
+			glog.Errorf("snapshot %v belongs to different volume %v", req.GetName(), snapInfo.ParentVolName)
+			return nil, status.Errorf(codes.AlreadyExists, "CreateSnapshot - snapshot %s belongs to different volume %s", snapInfo.ParentVolName, req.GetSourceVolumeId())
+		}
+
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				Id:             snapInfo.VolInfo.Name,
+				SourceVolumeId: snapInfo.ParentVolName,
+				CreatedAt:      snapInfo.CreatedAt.Unix(),
+				SizeBytes:      (int64(snapInfo.VolInfo.Capacity)) * utils.MB,
+				Status: &csi.SnapshotStatus{
+					Type: csi.SnapshotStatus_READY,
+				},
+			},
+		}, nil
+	}
+
+	snapReq := api.SnapCreateReq{
+		VolName:  req.SourceVolumeId,
+		SnapName: req.Name,
+		Force:    true,
+	}
+	glog.V(2).Infof("snapshot request: %+v", snapReq)
```

Comment:
This has to be hided in general

Reply:
volume create request is also logged with `V(2)` flag
```
glog.V(2).Infof("volume create request: %+v", volumeReq)
```